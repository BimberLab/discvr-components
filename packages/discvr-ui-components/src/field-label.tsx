import React from 'react';
import Select from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';


interface FieldModel {
    name: string;
    label: string | undefined;
    description: string;
    type: string;
    isInDefaultColumns: boolean;
    isMultiValued: boolean;
    isHidden: boolean;
    colWidth: number;
    formatString: string;
    orderKey: number;
    allowableValues: string[];
    category: string;
    url: string;
    flex: number;
    supportsFilter: boolean;
}


interface FilterType {
    field: string;
    operator: string;
    value: string;
}


type HandleFilterChangeFunction = (
    index: number, 
    key: string, 
    value: string | undefined, 
) => FilterType;


interface FieldLabelProps {
    filter: FilterType;
    index: number;
    handleFilterChange: HandleFilterChangeFunction;
    fieldTypeInfo: FieldModel[];
}


function FieldLabel(props: FieldLabelProps): React.ReactElement | null {

    const {filter, index, handleFilterChange, fieldTypeInfo} = props

    return (
        <Select
            label='Field'
            labelId="field-label"
            onChange={(event) =>
                handleFilterChange(index, "field", event.target.value)
            }
            value={filter.field}
        >
            <MenuItem style={{ display: 'none' }} value="">
                <em>None</em>
            </MenuItem>
            {fieldTypeInfo.map((field) => (
                <MenuItem key={field.name} value={field.name}>
                    {field.label ?? field.name}
                </MenuItem>
            ))}
        </Select>
    );

}

export {FieldLabel}