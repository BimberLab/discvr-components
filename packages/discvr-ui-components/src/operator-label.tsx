import React from 'react';
import Select from '@mui/material/Select';
import MenuItem from '@mui/material/MenuItem';

interface FieldModel {
    name: string;
    label: string | undefined;
    description: string;
    type: string;
    isInDefaultColumns: boolean;
    isMultiValued: boolean;
    isHidden: boolean;
    colWidth: number;
    formatString: string;
    orderKey: number;
    allowableValues: string[];
    category: string;
    url: string;
    flex: number;
    supportsFilter: boolean;
}

interface FilterType {
    field: string;
    operator: string;
    value: string;
}


type HandleFilterChangeFunction = (
    index: number,
    key: string,
    value: string | undefined,
) => FilterType;


type GetOperatorsForFieldFunction = (
    fieldTypeInfo: FieldModel | undefined
) => string[] | undefined


interface OperatorLabelProps {
    filter: FilterType;
    index: number;
    handleFilterChange: HandleFilterChangeFunction;
    fieldTypeInfo: FieldModel[];
    getOperatorsForField: GetOperatorsForFieldFunction;
}

function OperatorLabel(props: OperatorLabelProps): React.ReactElement | null {

    const {filter, index, getOperatorsForField, handleFilterChange, fieldTypeInfo} = props


    return (
        <Select
            label="Operator"    
            labelId="operator-label"

            onChange={(event) =>
                handleFilterChange(index, "operator", event.target.value)
            }
            value={filter.operator}
        >
            <MenuItem style={{ display: 'none' }} value="None">
                <em>None</em>
            </MenuItem>

            {getOperatorsForField(fieldTypeInfo.find(obj => obj.name === filter.field)) ? (
                getOperatorsForField(fieldTypeInfo.find(obj => obj.name === filter.field))?.map((operator) => (
                    <MenuItem key={operator} value={operator}>
                        {operator}
                    </MenuItem>
                ))
            ) : (
                <MenuItem/>
            )}

        </Select>
    );
}

export {OperatorLabel}