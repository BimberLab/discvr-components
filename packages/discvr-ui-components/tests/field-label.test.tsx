import React from 'react';
import {create} from 'react-test-renderer';
import { FieldLabel } from '../src/field-label';


export declare interface FieldModel {
    name: string
    label: string | undefined
    description: string
    type: string
    isInDefaultColumns: boolean
    isMultiValued: boolean
    isHidden: boolean
    colWidth: number
    formatString: string
    orderKey: number
    allowableValues: string[]
    category: string
    url: string
    flex: number
    supportsFilter: boolean
}

export declare interface FilterModel {
    field: string;
    operator: string;
    value: string;
}

it('FieldLabel test', () => {
    const fieldTypeInfo:FieldModel[] = [{
        name: '',
        label: '',
        description: '',
        type: '',
        isInDefaultColumns: false,
        isMultiValued: false,
        isHidden: false,
        colWidth: 0,
        formatString: '',
        orderKey: 0,
        allowableValues: [],
        category: '',
        url: '',
        flex: 0,
        supportsFilter: false
    },
    {
        name: '',
        label: '',
        description: '',
        type: '',
        isInDefaultColumns: false,
        isMultiValued: false,
        isHidden: false,
        colWidth: 0,
        formatString: '',
        orderKey: 0,
        allowableValues: [],
        category: '',
        url: '',
        flex: 0,
        supportsFilter: false
    }]

    const filter:FilterModel = {
        field: '',
        operator: '',
        value: ''
    }

    const component = create(
        <FieldLabel 
        fieldTypeInfo={fieldTypeInfo} 
        filter={filter} 
        handleFilterChange={(_index: number, _key: string, _value: string | undefined) => ({ field: "", operator: "", value: "" })}
        index={0}
        />
    );
    expect(component.toJSON()).toMatchSnapshot();
})