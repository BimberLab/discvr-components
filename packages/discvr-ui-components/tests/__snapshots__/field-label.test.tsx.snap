// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FieldLabel test 1`] = `
<div
  className="MuiInputBase-root MuiOutlinedInput-root MuiInputBase-colorPrimary  css-1yk1gt9-MuiInputBase-root-MuiOutlinedInput-root-MuiSelect-root"
  onClick={[Function]}
>
  <div
    aria-expanded="false"
    aria-haspopup="listbox"
    aria-labelledby="field-label"
    className="MuiSelect-select MuiSelect-outlined MuiInputBase-input MuiOutlinedInput-input css-11u53oe-MuiSelect-select-MuiInputBase-input-MuiOutlinedInput-input"
    onBlur={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onMouseDown={[Function]}
    role="combobox"
    tabIndex={0}
  >
    <span
      className="notranslate"
    >
      â€‹
    </span>
  </div>
  <input
    aria-hidden={true}
    aria-invalid={false}
    className="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
    onAnimationStart={[Function]}
    onChange={[Function]}
    tabIndex={-1}
    value=""
  />
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconOutlined css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
    data-testid="ArrowDropDownIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="M7 10l5 5 5-5z"
    />
  </svg>
  <fieldset
    aria-hidden={true}
    className="MuiOutlinedInput-notchedOutline css-1d3z3hw-MuiOutlinedInput-notchedOutline"
  >
    <legend
      className="css-yjsfm1"
    >
      <span>
        Field
      </span>
    </legend>
  </fieldset>
</div>
`;
